[Gradient Compression] Warm-start of PowerSGD

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49715 [Gradient Compression] Directly let world_size = process_group.size()
* #49711 [Gradient Compression] Explicitly restrict the scope of torch.cuda.synchronize to the current device
* #49709 [Gradient Compression] Change wait() to value() in some callbacks of PowerSGD communication hook
* **#49451 [Gradient Compression] Warm-start of PowerSGD**

Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

Commit:
[Gradient Compression] Warm-start of PowerSGD

Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -3781,15 +3782,22 @@ def _test_powerSGD_ddp_comm_hook_nccl(self, gradient_as_bucket_view=False):
-        )
+        # Test the hook with different algorithmic configs.
+        for use_error_feedback in [True, False]:
+            for warm_start in [True, False]:
```
nit: `for use_error_feedback, warm_start in itertools.product([T, F], [T, F])` is a bit more concise

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -367,56 +406,98 @@ def batched_powerSGD_hook(
+        state.q_memory_dict[bucket_index] = create_low_rank_tensor(
+            fill_random_values=True, rng=state.rng
+        )
+    _orthogonalize(state.q_memory_dict[bucket_index], 0)
```
Maybe I'm missing something, but feels like there is a slight difference between batched vs non-batched powersgd. 

in unbatched, we set p_memory_dict and q_memory_dict to torch.empty, and then later assign q.data = torch.rand, orthogonalize, and let p = TQ  for some tensor T. In this batched version, we instead initialize p_memory as empty, and q_memory as rand. We then orthogonalize q and let p = TQ. 

So the end result is the same, but it looks like unbatched version takes an extra step to assign q as empty but then re-assigns it to torch.rand. Can we skip that in the unbatched version?

SciPioneer replied:
Good observation! The subtle difference is true. This is actually intended behavior:
For the first un-batched version, need to call `torch.empty` to allocate contiguous memory (Qs) first, and then call `torch.rand` to randomize each individual Q. Cannot directly randomize Qs at one time like the batched version.

rohan-varma reviewed (approved):
LGTM. I suppose that we'll follow up with information regarding data around the accuracy/speed tradeoffs for all of these configurations? Would be great to get some data around how the different flags trade off.

SciPioneer commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -174,16 +186,35 @@ def powerSGD_hook(
+        # If warm-start is disabled, low-rank tensors will be initialized at every step.
+        # Only log this if warm-start to avoid spamming.
+        if state.warm_start:
+            logging.info(
```
I think we should use the same log level as ["Reducer buckets have been rebuilt in this iteration."](https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L672), as this only happens when the warm-start is enabled and the buckets are rebuilt.

SciPioneer commented:
> LGTM. I suppose that we'll follow up with information regarding data around the accuracy/speed tradeoffs for all of these configurations? Would be great to get some data around how the different flags trade off.

Have collected some experimental results for only one experiment so far, but still need experiment on a more training cases to draw a conclusion.

Commit:
Update on "[Gradient Compression] Warm-start of PowerSGD"


Reuse the low-rank tensors P(s) and Q(s) from the previous iteration if possible.

This can give a better compression performance in terms of both accuracy and speed.

Also add a unit test for batched PowerSGD to test_c10d.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25583086](https://our.internmc.facebook.com/intern/diff/D25583086/)

[ghstack-poisoned]
